pwiz_tools/Shared/PanoramaClient/PanoramaClient.cs

```diff
@@ -0,0 +1,762 @@
+
+            return tmpUploadUri;
+        }
+
+        private string GetWebDavPath(string folderPath, IRequestHelper requestHelper)
+        {
+            var getPipelineContainerUri = PanoramaUtil.GetPipelineContainerUrl(ServerUri, folderPath);
+
+            var jsonResponse = requestHelper.Get(getPipelineContainerUri,
+                string.Format(
+                    Resources
+                        .AbstractPanoramaClient_GetWebDavPath_There_was_an_error_getting_the_WebDAV_url_for_folder___0__,
+                    folderPath));
+            var webDavUrl = (string)jsonResponse[@"webDavURL"];
+            if (webDavUrl == null)
+            {
+                throw new PanoramaServerException(new ErrorMessageBuilder(Resources.AbstractPanoramaClient_GetWebDavPath_Missing_webDavURL_in_response_)
+                    .Uri(getPipelineContainerUri).Response(jsonResponse).Build());
+            }
+            return webDavUrl;
+        }
+
+        private class ImportStatus
+        {
+            public string StatusString { get; }
+            public bool IsComplete => string.Equals(@"COMPLETE", StatusString);
+            public bool IsRunning => StatusString.Contains(@"RUNNING"); // "IMPORT RUNNING" pre LK19.3, RUNNING, x% in LK19.3
+            public bool IsError => string.Equals(@"ERROR", StatusString);
+            public bool IsCancelled => string.Equals(@"CANCELLED", StatusString);
+
+            public ImportStatus(string status)
+            {
+                StatusString = status;
+            }
+        }
+
+        private void updateProgressAndWait(ImportStatus jobStatus, IProgressMonitor progressMonitor, DateTime startTime)
+        {
+            var match = _runningStatusRegex.Match(jobStatus.StatusString);
+            if (match.Success)
+            {
+                var currentProgress = _progressStatus.PercentComplete;
+
+                if (int.TryParse(match.Groups[1].Value, out var progress))
+                {
+                    progress = Math.Max(progress, currentProgress);
+                    _progressStatus = _progressStatus.ChangeMessage(string.Format(
+                        Resources.AbstractPanoramaClient_updateProgressAndWait_Importing_data___0___complete_,
+                        progress));
+                    _progressMonitor.UpdateProgress(_progressStatus = _progressStatus.ChangePercentComplete(progress));
+
+                    var delta = progress - currentProgress;
+                    if (delta > 1)
+                    {
+                        // If progress is > 1% half the wait time
+                        _waitTime = Math.Max(1, _waitTime / 2);
+                    }
+                    else if (delta < 1)
+                    {
+                        // If progress is < 1% double the wait time, up to a max of 10 seconds.
+                        _waitTime = Math.Min(10, _waitTime * 2);
+                    }
+
+                    Thread.Sleep(_waitTime * 1000);
```

Comment:
It looks like this gets called from "LongWaitDlg.RunWork".
If the user hits the Cancel button in the LongWaitDlg, the dialog stays up until the thread it launched finishes.
For this reason, instead of doing Thread.Sleep(_waitTime * 1000), it would be better to do:
```
for (int i = 0; i < _waitTime * 10; i++) {
    if (progressMonitor.IsCancelled)
         return;
    Thread.Sleep(100);
}
```
so that the user never has to wait more than 100msec for the LongWaitDlg to disappear.

Reply:
Thanks!